### PR TITLE
Extracting Store into multiple stores

### PR DESCRIPTION
**Pull request by vrackovic**

### CODE REVIEW

File: web/src/main/frontend/src/base/stores/Category.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  category() {
+    return util.splitTableDataHeadings(
+      this.categories.map(category => category)
+    );
+  }
+
+  @computed
+  get categoryTable() {
+    return util.tableHandler(this.category());
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/Category.js]: Let's install Ramda (http://ramdajs.com/) and make use of `compose`. 😄

Comment (iscekic) [reply web/src/main/frontend/src/base/stores/Category.js]: Same in the other stores.

File: web/src/main/frontend/src/base/stores/Category.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import {computed, observable} from 'mobx';
+import * as util from '../util.js';
+
+class Category {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/Category.js]: Rename to `CategoryStore`.

Same for the other stores.

File: web/src/main/frontend/src/base/util.js [hunk 3]
```diff
@@ -1,15 +1,10 @@
...
-});
+export const addControls = obj => ({...obj, controls: true});
 
 // Trims 'id' from tableHeader
 export const tableHeaderRows = obj => obj.filter(data => data !== 'id');
 
-// tableHeadings is used for computing different table header fields from matching observable keys
-export const tableHeadings = table => ({
+// Spliting table data with headings
+export const splitTableDataHeadings = table => ({
```
Comment (iscekic) [inline web/src/main/frontend/src/base/util.js]: Recheck function names in this file and make sure they are good.

File: web/src/main/frontend/src/base/stores/CategoryStore.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    {id: 2, title: 'Web', text: 'Questions for Web developers'},
+    {id: 3, title: 'Android', text: 'Questions for Android developers'},
+    {id: 4, title: 'Java', text: 'Questions for Java developers'},
+    {id: 5, title: 'Web', text: 'Questions for Web developers'},
+    {id: 6, title: 'Android', text: 'Questions for Android developers'}
+  ];
+
+  @computed
+  get categoryTable() {
+    return R.compose(util.tableTransformer, util.splitDataAndHeadings)(
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/CategoryStore.js]: Extract `R.compose(util.tableTransformer, util.splitDataAndHeadings)` to a variable.

File: web/src/main/frontend/src/base/stores/CategoryStore.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import {computed, observable} from 'mobx';
+import * as util from '../util.js';
+import * as R from 'ramda';
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/CategoryStore.js]: Don't import `*`. Only import what you need. eg:

```js
import compose from 'ramda/compose';
// or
import { compose } from 'ramda';
```

File: web/src/main/frontend/src/base/stores/CategoryStore.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    {id: 3, title: 'Android', text: 'Questions for Android developers'},
+    {id: 4, title: 'Java', text: 'Questions for Java developers'},
+    {id: 5, title: 'Web', text: 'Questions for Web developers'},
+    {id: 6, title: 'Android', text: 'Questions for Android developers'}
+  ];
+
+  @computed
+  get categoryTable() {
+    return R.compose(util.tableTransformer, util.splitDataAndHeadings)(
+      this.categories.slice()
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/CategoryStore.js]: Why `slice` here?

File: web/src/main/frontend/src/base/stores/CategoryStore.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    );
+  }
+
+  @computed
+  get categorySelections() {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/CategoryStore.js]: What are `categorySelections`?

File: web/src/main/frontend/src/base/stores/QuestionStore.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+      id: 1,
+      question: 'What is the difference between throw and throws',
+      points: '15'
+    }
+  ];
+
+  @computed
+  get questionTable() {
+    return R.compose(util.tableTransformer, util.splitDataAndHeadings)(
+      this.questions.slice()
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/QuestionStore.js]: Why `slice` here?

File: web/src/main/frontend/src/base/util.js [hunk 9]
```diff
@@ -1,30 +1,28 @@
-export const addControls = obj => ({...obj, controls: ['edit', 'delete']});
-
-export const addRandomId = obj => ({
-  ...obj,
-  id: Math.floor(Math.random() * 100000)
-});
+export const addControlsField = obj =>
+  obj.map(row => ({...row, controls: true}));
```
Comment (iscekic) [inline web/src/main/frontend/src/base/util.js]: What is `controls` used for?

File: web/src/main/frontend/src/base/util.js [hunk 10]
```diff
@@ -1,30 +1,28 @@
...
+// ({...obj, controls: true});
 
 // Trims 'id' from tableHeader
-export const tableHeaderRows = obj => obj.filter(data => data !== 'id');
+export const trimIdField = obj => obj.filter(data => data !== 'id');
```
Comment (iscekic) [inline web/src/main/frontend/src/base/util.js]: Instead of having multiple `trim*` functions, make a single `filterOut`, and then compose where necessary.

File: web/src/main/frontend/src/base/util.js [hunk 11]
```diff
@@ -1,30 +1,28 @@
...
+
+// Trims 'id' and 'controls' from tableData
+export const trimIdAndControlsFields = obj =>
+  obj.filter(heading => heading !== 'controls' && heading !== 'id');
 
-// tableHeadings is used for computing different table header fields from matching observable keys
-export const tableHeadings = table => ({
+// Spliting table data with headings
+export const splitDataAndHeadings = table => ({
```
Comment (iscekic) [inline web/src/main/frontend/src/base/util.js]: Not sure I understand this fn. `table` seems to be an array of objects, then you spread it into an object? Split this one into multiple smaller functions and compose them.

File: web/src/main/frontend/src/base/util.js [hunk 12]
```diff
@@ -1,30 +1,28 @@
...
   tableHeadings: Object.keys(table[0])
 });
 
-// Trims 'id' and 'controls' from tableData
-export const tableHeadingsFiltered = obj =>
-  obj.filter(heading => heading !== 'controls' && heading !== 'id');
-
 // Converts table object into array containing objects that fit into <tr></tr>
-export const tableRows = obj => Object.keys(obj).map(key => obj[key]);
+export const tableObjectToArray = obj => Object.keys(obj).map(key => obj[key]);
```
Comment (iscekic) [inline web/src/main/frontend/src/base/util.js]: Use `R.values`.

File: web/src/main/frontend/src/base/stores/CategoryStore.js (hunk 7 above)
Comment (vrackovic) [reply web/src/main/frontend/src/base/stores/CategoryStore.js]: Creates array of titles. It is used in other components for loading html options element, and select elements are passed from this title array.

File: web/src/main/frontend/src/base/stores/CategoryStore.js (hunk 6 above)
Comment (vrackovic) [reply web/src/main/frontend/src/base/stores/CategoryStore.js]: To pass over data without wrapping observable.

File: web/src/main/frontend/src/base/stores/QuestionStore.js (hunk 8 above)
Comment (vrackovic) [reply web/src/main/frontend/src/base/stores/QuestionStore.js]: To pass over data without wrapping observable

File: web/src/main/frontend/src/base/util.js (hunk 9 above)
Comment (vrackovic) [reply web/src/main/frontend/src/base/util.js]: To display `controls` field that contains `edit` and `delete` controls in table in case user has `admin` role

File: web/src/main/frontend/src/base/stores/CategoryStore.js [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    {id: 2, title: 'Web', text: 'Questions for Web developers'},
+    {id: 3, title: 'Android', text: 'Questions for Android developers'},
+    {id: 4, title: 'Java', text: 'Questions for Java developers'},
+    {id: 5, title: 'Web', text: 'Questions for Web developers'},
+    {id: 6, title: 'Android', text: 'Questions for Android developers'}
+  ];
+
+  @computed
+  get categoryTable() {
+    return util.tableTransformer(this.categories.map(category => category));
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/CategoryStore.js]: Why do we have `category => category` here again?

Comment (iscekic) [reply web/src/main/frontend/src/base/stores/CategoryStore.js]: If you need to create a copy, just use `[...this.categories]`.

File: web/src/main/frontend/src/base/stores/DashboardStore.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+      test: 'C#',
+      date_taken: '03.10.2017',
+      max_points: 120,
+      points: 100
+    }
+  ];
+
+  @computed
+  get dashboardTable() {
+    const transformTable = compose(
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/DashboardStore.js]: This variable can be extracted above the class, or as a class property. No need to initialize it each time. 😄

File: web/src/main/frontend/src/base/stores/FormulaStore.js [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    {
+      id: 5,
+      formula: 'Android Test',
+      description: 'Test containing...'
+    }
+  ];
+
+  @computed
+  get formulaTable() {
+    const transformTable = compose(util.tableTransformer);
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/FormulaStore.js]: No need to compose a single function.

File: web/src/main/frontend/src/base/stores/FormulaStore.js [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    return transformTable(this.formulas.map(formula => formula));
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/FormulaStore.js]: Same as above.

File: web/src/main/frontend/src/base/stores/QuestionStore.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+import {computed, observable} from 'mobx';
+import * as util from '../util.js';
+import * as R from 'ramda';
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/QuestionStore.js]: Don't import `*`.

File: web/src/main/frontend/src/base/stores/QuestionStore.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+    {
+      id: 1,
+      question: 'What is the difference between throw and throws',
+      points: '15'
+    }
+  ];
+
+  @computed
+  get questionTable() {
+    const transformTable = R.compose(util.tableTransformer);
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/QuestionStore.js]: Same as above.

File: web/src/main/frontend/src/base/stores/QuestionStore.js [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+    return transformTable(this.questions.map(question => question));
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/QuestionStore.js]: Same as above.

File: web/src/main/frontend/src/base/util.js [hunk 20]
```diff
@@ -1,30 +1,25 @@
...
-  id: Math.floor(Math.random() * 100000)
-});
+// Add 'controls' field to table if user has admin privileges
+export const addControlsField = obj =>
+  obj.map(row => ({...row, controls: true}));
 
-// Trims 'id' from tableHeader
-export const tableHeaderRows = obj => obj.filter(data => data !== 'id');
+// Trims key from array passed with parameter
+export const filterOut = (obj, keyToFilter) =>
```
Comment (iscekic) [inline web/src/main/frontend/src/base/util.js]: This function can be written as a composition of [filter](http://ramdajs.com/docs/#filter) and [has](http://ramdajs.com/docs/#has).

File: web/src/main/frontend/src/base/util.js [hunk 21]
```diff
@@ -1,30 +1,25 @@
...
+  obj.filter(heading => heading !== keyToFilter);
 
-// tableHeadings is used for computing different table header fields from matching observable keys
-export const tableHeadings = table => ({
-  tableData: {...table},
-  tableHeadings: Object.keys(table[0])
-});
+// Gets table as parameter and returns first table rows keys
+export const tableKeys = table => keys(table[0]);
```
Comment (iscekic) [inline web/src/main/frontend/src/base/util.js]: This is a copy of http://ramdajs.com/docs/#keys

File: web/src/main/frontend/src/base/util.js [hunk 22]
```diff
@@ -1,30 +1,29 @@
...
-  id: Math.floor(Math.random() * 100000)
-});
+// Add 'controls' field to table if user has admin privileges
+export const addControlsField = obj =>
+  obj.map(row => ({...row, controls: true}));
 
-// Trims 'id' from tableHeader
-export const tableHeaderRows = obj => obj.filter(data => data !== 'id');
+// Gets table as parameter and returns first table rows keys
+export const firstItemKeys = table => keys(head(table));
```
Comment (iscekic) [inline web/src/main/frontend/src/base/util.js]: Can be written as:

```js
export const firstItemKeys = compose(keys, head);
```

File: web/src/main/frontend/src/base/util.js [hunk 23]
```diff
@@ -1,30 +1,29 @@
...
-    headerRows: tableHeaderRows(table.tableHeadings)
+    headingsFiltered: filteredTableControls,
+    tableRows: table,
+    headerRows: filteredTableId
   };
 };
+
+// Gets array of function calls passed as param and returns composition of those funtions.
+// Call odred matters
+export const composeMultiple = funcArr => {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/util.js]: No need for this, just use `compose` directly.

File: web/src/main/frontend/src/base/stores/CategoryStore.js [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+    {id: 6, title: 'Android', text: 'Questions for Android developers'}
+  ];
+
+  @computed
+  get categoryTable() {
+    return util.tableTransformer([...this.categories]);
+  }
+
+  @computed
+  get categorySelections() {
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/CategoryStore.js]: Rename to `categoryTitles`.

File: web/src/main/frontend/src/base/stores/DashboardStore.js [hunk 25]
```diff
@@ -0,0 +1,55 @@
+import {computed, observable} from 'mobx';
+import * as util from '../util.js';
+import {compose} from 'ramda';
+
+class DashboardStore {
+  t;
```
Comment (iscekic) [inline web/src/main/frontend/src/base/stores/DashboardStore.js]: Why is this here?

Comment (vrackovic) [reply web/src/main/frontend/src/base/stores/DashboardStore.js]: I have no idea :smile:
